tests/integration/refresh/test_refresh.py

```diff
@@ -0,0 +1,279 @@
+    """
+    username = "admin"
+    password = fetch_password(juju)
+    assert validate_sql_queries_with_kyuubi(
+        juju=juju,
+        db_name=DB_NAME,
+        table_name=TABLE_NAME,
+        username=username,
+        password=password,
+        use_tls=with_tls,
+    )
+
+
+def test_pre_refresh_check(juju: jubilant.Juju) -> None:
+    """Test that the pre-refresh-check action runs successfully."""
+    logger.info("Get leader unit")
+    leader_unit = get_leader_unit(juju, APP_NAME)
+
+    logger.info("Run pre-refresh-check action")
+    task = juju.run(leader_unit, "pre-refresh-check")
+    assert task.return_code == 0
+
+
+def test_run_inplace_upgrade(
+    juju: jubilant.Juju,
+    kyuubi_charm: Path,
+    with_image_upgrade: bool,
+    with_multi_units: bool,
+) -> None:
+    """Test that the inplace upgrade leads to an active deployment."""
+    refresh_args: dict = {"path": str(kyuubi_charm)}
+    if with_image_upgrade:
+        refresh_args["resources"] = {"kyuubi-image": WORKLOAD_IMAGE_UPGRADE}
+        logger.info(f"Will upgrade workload image to: {WORKLOAD_IMAGE_UPGRADE}")
+
+    logger.info("Refreshing Kyuubi")
+
+    # start refresh by upgrading to the current version
+    juju.refresh(APP_NAME, **refresh_args)
+
+    logger.info("Waiting for upgrade to start")
+
+    if not with_multi_units:
+        # fast track, one unit will not block in any case
+        juju.wait(jubilant.all_active, delay=10)
+        return
+
+    status = juju.wait(lambda status: jubilant.all_agents_idle(status, APP_NAME), delay=10)
+    # Highest to lowest unit number
+    refresh_order = sorted(
+        status.apps[APP_NAME].units.keys(),
+        key=lambda unit: int(unit.split("/")[1]),
+        reverse=True,
+    )
+
+    task_params = {"check-compatibility": False}
+
+    if not with_image_upgrade:
+        assert (
+            "Refresh incompatible"
+            in status.apps[APP_NAME].units[refresh_order[0]].workload_status.message
+        ), "Application refresh not blocked due to incompatibility"
+
+    else:
+        assert (
+            "missing/incorrect OCI resource"
+            in status.apps[APP_NAME].units[refresh_order[0]].workload_status.message
+        ), "Application refresh not blocked due to image incompatibility"
+        task_params["check-workload-container"] = False
+
+    task = juju.run(refresh_order[0], "force-refresh-start", task_params)
+    assert task.return_code == 0
+
+    logger.info("Waiting for first unit to upgrade")
+    status = juju.wait(lambda status: jubilant.all_agents_idle(status, APP_NAME), delay=10)
+
+    logger.info("Running resume-refresh action")
+    leader_unit = get_leader_unit(juju, APP_NAME)
+    try:
+        juju.run(leader_unit, "resume-refresh")
+    except jubilant.TaskError:
+        # TODO: Ask about this
+        # Not sure why we get a task terminated while things are apparently going fine.
```

Comment:
This is a juju bug, see https://github.com/canonical/charm-refresh/blob/4c82e341084b50443144511b1ca40218bcaa6165/charm_refresh/_main.py#L1872-L1874.

**todo**: document here